File: src/git.ts
import { execFileSync } from 'child_process';

type ToolResult = { content: Array<{ type: 'text'; text: string }> };

const JIRA_KEY_RE = /\b[A-Z][A-Z0-9]+-\d+\b/g;

// Allowlist for git refs (commits, branches used as refs in diff commands)
const SAFE_REF_RE = /^[a-zA-Z0-9/_.\-@{}~^:]+(\.\.\.[a-zA-Z0-9/_.\-@{}~^:]+)?$/;
// Allowlist for branch names (stricter — no range syntax)
const SAFE_BRANCH_RE = /^[a-zA-Z0-9/_.\-]+$/;

function text(t: string): ToolResult {
  return { content: [{ type: 'text', text: t }] };
}

function git(args: string[], cwd: string): string {
  return execFileSync('git', args, { cwd, encoding: 'utf-8', stdio: ['pipe', 'pipe', 'pipe'] }).trim();
}

function safeGit(args: string[], cwd: string, fallback = ''): string {
  try {
    return git(args, cwd);
  } catch {
    return fallback;
  }
}

function validateRepoPath(repoPath: string): void {
  try {
    execFileSync('git', ['rev-parse', '--git-dir'], { cwd: repoPath, encoding: 'utf-8', stdio: 'pipe' });
  } catch {
    throw new Error(`Not a git repository: ${repoPath}`);
  }
}

function validateBranch(branch: string, label: string): void {
  if (!SAFE_BRANCH_RE.test(branch)) {
    throw new Error(`Invalid ${label} "${branch}". Use only letters, numbers, /, _, ., -`);
  }
}

function validateRef(ref: string, label: string): void {
  if (!SAFE_REF_RE.test(ref)) {
    throw new Error(`Invalid ${label} "${ref}". Use only safe git ref characters.`);
  }
}

export function getContext(args: { repoPath?: string; commitLimit?: number; includeDiff?: boolean }): ToolResult {
  const repoPath = args.repoPath ?? process.cwd();
  const limit = Math.max(1, Math.min(args.commitLimit ?? 10, 100));
  try {
    validateRepoPath(repoPath);
    const branch = safeGit(['rev-parse', '--abbrev-ref', 'HEAD'], repoPath, '(unknown)');
    const remote = safeGit(['remote', 'get-url', 'origin'], repoPath, '(no remote)');
    const commits = safeGit(['log', '--oneline', `-${limit}`], repoPath, '(no commits)');
    const status = safeGit(['status', '--short'], repoPath, '');

    // Upstream tracking
    const upstream = safeGit(['rev-parse', '--abbrev-ref', '@{u}'], repoPath, '');
    let upstreamLine = '';
    if (upstream) {
      const ab = safeGit(['rev-list', '--left-right', '--count', `${upstream}...HEAD`], repoPath, '');
      if (ab.includes('\t')) {
        const [behind, ahead] = ab.split('\t').map(Number);
        const parts: string[] = [];
        if (ahead) parts.push(`${ahead} ahead`);
        if (behind) parts.push(`${behind} behind`);
        upstreamLine = `${upstream}${parts.length ? ` (${parts.join(', ')})` : ' (up to date)'}`;
      }
    }

    // Diff stat summary
    const diffStatLines = safeGit(['diff', 'HEAD', '--stat'], repoPath, '').split('\n').filter(Boolean);
    const diffStat = diffStatLines[diffStatLines.length - 1]?.trim() ?? '';

    const jiraKeys = [...new Set(branch.match(JIRA_KEY_RE) ?? [])];

    const lines = [
      `Repository: ${repoPath}`,
      `Branch:     ${branch}`,
      ...(upstreamLine ? [`Upstream:   ${upstreamLine}`] : []),
      `Remote:     ${remote}`,
      ...(jiraKeys.length ? [`Jira:       ${jiraKeys.join(', ')}`] : []),
      '',
      `Recent commits (last ${limit}):`,
      commits || '(none)',
      '',
      'Working tree:',
    ];

    if (status) {
      lines.push(status);
      if (diffStat) lines.push('', `Diff stat:  ${diffStat}`);
    } else {
      lines.push('(clean)');
    }

    if (args.includeDiff && status) {
      const diff = safeGit(['diff', 'HEAD'], repoPath, '');
      if (diff) {
        const MAX = 6000;
        lines.push(
          '',
          '── Uncommitted diff ──',
          diff.length > MAX ? diff.slice(0, MAX) + `\n\n... (truncated, ${diff.length - MAX} more chars)` : diff,
        );
      }
    }

    return text(lines.join('\n'));
  } catch (err) {
    return text(`Error reading git context: ${(err as Error).message}`);
  }
}

// Not exposed as an MCP tool — internal/experimental use only
export function getCommits(args: {
  repoPath?: string;
  limit?: number;
  branch?: string;
}): ToolResult {
  const repoPath = args.repoPath ?? process.cwd();
  const limit = Math.max(1, Math.min(args.limit ?? 20, 200));
  const branch = args.branch ?? '';
  try {
    validateRepoPath(repoPath);
    if (branch) validateBranch(branch, 'branch');
    const format = '%H|%an|%ad|%s';
    const gitArgs = ['log', `--pretty=format:${format}`, '--date=short', `-${limit}`, ...(branch ? [branch] : [])];
    const raw = safeGit(gitArgs, repoPath, '');
    if (!raw) return text('No commits found.');
    const lines = raw.split('\n').map((line) => {
      const [hash, author, date, ...msgParts] = line.split('|');
      return `${hash?.slice(0, 8)} ${date} ${author}: ${msgParts.join('|')}`;
    });
    return text(`Last ${lines.length} commit(s)${branch ? ` on ${branch}` : ''}:\n${lines.join('\n')}`);
  } catch (err) {
    return text(`Error reading commits: ${(err as Error).message}`);
  }
}

export function getDiff(args: {
  repoPath?: string;
  fromRef?: string;
  toRef?: string;
}): ToolResult {
  const repoPath = args.repoPath ?? process.cwd();
  try {
    validateRepoPath(repoPath);
    let gitArgs: string[];
    if (args.fromRef && args.toRef) {
      validateRef(args.fromRef, 'fromRef');
      validateRef(args.toRef, 'toRef');
      gitArgs = ['diff', args.fromRef, args.toRef];
    } else if (args.fromRef) {
      validateRef(args.fromRef, 'fromRef');
      gitArgs = ['diff', args.fromRef];
    } else {
      gitArgs = ['diff', 'HEAD'];
    }
    const diff = safeGit(gitArgs, repoPath, '');
    if (!diff) return text('No differences found.');
    return text(diff);
  } catch (err) {
    return text(`Error reading diff: ${(err as Error).message}`);
  }
}

export function checkRemoteBranch(branchName: string, repoPath: string): {
  exists: boolean; author?: string; date?: string; message?: string; sha?: string;
} {
  validateBranch(branchName, 'branchName');
  const lsRemote = safeGit(['ls-remote', '--heads', 'origin', `refs/heads/${branchName}`], repoPath);
  if (!lsRemote) return { exists: false };
  const sha = lsRemote.split(/\s+/)[0]?.trim();
  // Fetch so we can read the log — failure is non-fatal (network/credentials issue)
  try {
    git(['fetch', 'origin', branchName], repoPath);
  } catch {
    return { exists: true, sha: sha?.slice(0, 8) };
  }
  const log = safeGit(['log', `origin/${branchName}`, '-1', '--format=%an%x09%ae%x09%ad%x09%s'], repoPath);
  if (!log) return { exists: true, sha: sha?.slice(0, 8) };
  const [author, email, date, ...msgParts] = log.split('\t');
  return {
    exists: true,
    sha: sha?.slice(0, 8),
    author: email ? `${author} <${email}>` : author,
    date,
    message: msgParts.join('\t'),
  };
}

function getDefaultBranch(repoPath: string): string {
  const head = safeGit(['rev-parse', '--abbrev-ref', 'origin/HEAD'], repoPath);
  if (head && head.startsWith('origin/')) return head.slice('origin/'.length);
  // origin/HEAD not set — probe common defaults
  if (safeGit(['rev-parse', '--verify', 'origin/main'], repoPath)) return 'main';
  return 'master';
}

export function checkoutRemoteBranch(branchName: string, repoPath: string): ToolResult {
  try {
    validateBranch(branchName, 'branchName');
    const existing = safeGit(['branch', '--list', branchName], repoPath);
    if (existing.trim()) {
      git(['checkout', branchName], repoPath);
      return text(`Switched to existing local branch "${branchName}".`);
    }
    git(['checkout', '--track', `origin/${branchName}`], repoPath);
    return text(`Checked out "${branchName}" tracking origin/${branchName}.`);
  } catch (err) {
    return text(`Error checking out branch: ${(err as Error).message}`);
  }
}

export function createBranch(args: {
  branchName: string;
  baseBranch?: string;
  repoPath?: string;
  push?: boolean;
}): ToolResult {
  const repoPath = args.repoPath ?? process.cwd();
  const { branchName, push = false } = args;
  try {
    validateRepoPath(repoPath);
    if (!SAFE_BRANCH_RE.test(branchName)) {
      return text(`Invalid branch name "${branchName}". Use only letters, numbers, /, _, ., -`);
    }
    const baseBranch = args.baseBranch ?? getDefaultBranch(repoPath);
    if (!SAFE_BRANCH_RE.test(baseBranch)) {
      return text(`Invalid base branch name "${baseBranch}". Use only letters, numbers, /, _, ., -`);
    }
    const existing = safeGit(['branch', '--list', branchName], repoPath);
    if (existing.trim()) {
      return text(`Branch "${branchName}" already exists locally. Switch with: git checkout ${branchName}`);
    }
    safeGit(['fetch', 'origin', baseBranch], repoPath);
    git(['checkout', '-b', branchName, `origin/${baseBranch}`], repoPath);
    const lines = [`Created and switched to branch "${branchName}" from origin/${baseBranch}.`];
    if (push) {
      git(['push', '-u', 'origin', branchName], repoPath);
      lines.push(`Pushed to origin/${branchName} and set upstream.`);
    }
    return text(lines.join('\n'));
  } catch (err) {
    return text(`Error creating branch: ${(err as Error).message}`);
  }
}

File: git.go
package main

import (
	"fmt"
	"net/url"
	"os/exec"
	"regexp"
	"strconv"
	"strings"
)

// Allowlist for git refs (commits, branches used as refs in diff commands).
var safeRefRe = regexp.MustCompile(`^[a-zA-Z0-9/_.\-@{}~^:]+(\.\.\.[a-zA-Z0-9/_.\-@{}~^:]+)?$`)

// Allowlist for branch names (stricter — no range syntax).
var safeBranchRe = regexp.MustCompile(`^[a-zA-Z0-9/_.\-]+$`)

// gitIn runs git in cwd and returns trimmed stdout, or an error.
func gitIn(cwd string, args ...string) (string, error) {
	cmd := exec.Command("git", args...)
	if cwd != "" {
		cmd.Dir = cwd
	}
	out, err := cmd.Output()
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(out)), nil
}

// safeGit runs git in cwd, returning fallback on any error.
func safeGit(cwd string, fallback string, args ...string) string {
	out, err := gitIn(cwd, args...)
	if err != nil {
		return fallback
	}
	return out
}

// safeExecGit runs git in the process cwd, returning "" on error.
func safeExecGit(args ...string) string {
	return safeGit("", "", args...)
}

func currentGitRemote() string {
	return safeExecGit("remote", "get-url", "origin")
}

func currentGitBranch() string {
	return safeExecGit("rev-parse", "--abbrev-ref", "HEAD")
}

func remoteMatchesBitbucketInstance(remote, bitbucketURL string) bool {
	if remote == "" {
		return false
	}
	u, err := url.Parse(bitbucketURL)
	if err != nil || u.Hostname() == "" {
		return false
	}
	return strings.Contains(strings.ToLower(remote), strings.ToLower(u.Hostname()))
}

func validateRepoPath(repoPath string) error {
	if _, err := gitIn(repoPath, "rev-parse", "--git-dir"); err != nil {
		return fmt.Errorf("Not a git repository: %s", repoPath)
	}
	return nil
}

func validateBranch(branch, label string) error {
	if !safeBranchRe.MatchString(branch) {
		return fmt.Errorf("Invalid %s %q. Use only letters, numbers, /, _, ., -", label, branch)
	}
	return nil
}

func validateRef(ref, label string) error {
	if !safeRefRe.MatchString(ref) {
		return fmt.Errorf("Invalid %s %q. Use only safe git ref characters.", label, ref)
	}
	return nil
}

func repoPathOrCwd(args map[string]any) string {
	if p := argString(args, "repoPath"); p != "" {
		return p
	}
	return mustGetwd()
}

// gitGetContext implements git_get_context.
func gitGetContext(args map[string]any) toolResult {
	repoPath := repoPathOrCwd(args)
	limit := argIntDefault(args, "commitLimit", 10)
	if limit < 1 {
		limit = 1
	} else if limit > 100 {
		limit = 100
	}
	if err := validateRepoPath(repoPath); err != nil {
		return textResult("Error reading git context: " + err.Error())
	}
	branch := safeGit(repoPath, "(unknown)", "rev-parse", "--abbrev-ref", "HEAD")
	remote := safeGit(repoPath, "(no remote)", "remote", "get-url", "origin")
	commits := safeGit(repoPath, "(no commits)", "log", "--oneline", fmt.Sprintf("-%d", limit))
	status := safeGit(repoPath, "", "status", "--short")

	upstream := safeGit(repoPath, "", "rev-parse", "--abbrev-ref", "@{u}")
	upstreamLine := ""
	if upstream != "" {
		ab := safeGit(repoPath, "", "rev-list", "--left-right", "--count", upstream+"...HEAD")
		if strings.Contains(ab, "\t") {
			parts := strings.SplitN(ab, "\t", 2)
			behind, _ := strconv.Atoi(strings.TrimSpace(parts[0]))
			ahead, _ := strconv.Atoi(strings.TrimSpace(parts[1]))
			var p []string
			if ahead != 0 {
				p = append(p, fmt.Sprintf("%d ahead", ahead))
			}
			if behind != 0 {
				p = append(p, fmt.Sprintf("%d behind", behind))
			}
			if len(p) > 0 {
				upstreamLine = upstream + " (" + strings.Join(p, ", ") + ")"
			} else {
				upstreamLine = upstream + " (up to date)"
			}
		}
	}

	diffStat := ""
	diffStatLines := splitNonEmpty(safeGit(repoPath, "", "diff", "HEAD", "--stat"), "\n")
	if len(diffStatLines) > 0 {
		diffStat = strings.TrimSpace(diffStatLines[len(diffStatLines)-1])
	}

	jiraKeys := uniqueStrings(jiraKeyRe.FindAllString(branch, -1))

	lines := []string{
		"Repository: " + repoPath,
		"Branch:     " + branch,
	}
	if upstreamLine != "" {
		lines = append(lines, "Upstream:   "+upstreamLine)
	}
	lines = append(lines, "Remote:     "+remote)
	if len(jiraKeys) > 0 {
		lines = append(lines, "Jira:       "+strings.Join(jiraKeys, ", "))
	}
	lines = append(lines, "", fmt.Sprintf("Recent commits (last %d):", limit), orValue(commits, "(none)"), "", "Working tree:")

	if status != "" {
		lines = append(lines, status)
		if diffStat != "" {
			lines = append(lines, "", "Diff stat:  "+diffStat)
		}
	} else {
		lines = append(lines, "(clean)")
	}

	if argBool(args, "includeDiff") && status != "" {
		diff := safeGit(repoPath, "", "diff", "HEAD")
		if diff != "" {
			const max = 6000
			body := diff
			if len(diff) > max {
				body = diff[:max] + fmt.Sprintf("\n\n... (truncated, %d more chars)", len(diff)-max)
			}
			lines = append(lines, "", "── Uncommitted diff ──", body)
		}
	}

	return textResult(strings.Join(lines, "\n"))
}

// getDiff implements the core of git_get_diff (before paging).
func getDiff(args map[string]any) toolResult {
	repoPath := repoPathOrCwd(args)
	if err := validateRepoPath(repoPath); err != nil {
		return textResult("Error reading diff: " + err.Error())
	}
	fromRef := argString(args, "fromRef")
	toRef := argString(args, "toRef")
	var gitArgs []string
	switch {
	case fromRef != "" && toRef != "":
		if err := validateRef(fromRef, "fromRef"); err != nil {
			return textResult("Error reading diff: " + err.Error())
		}
		if err := validateRef(toRef, "toRef"); err != nil {
			return textResult("Error reading diff: " + err.Error())
		}
		gitArgs = []string{"diff", fromRef, toRef}
	case fromRef != "":
		if err := validateRef(fromRef, "fromRef"); err != nil {
			return textResult("Error reading diff: " + err.Error())
		}
		gitArgs = []string{"diff", fromRef}
	default:
		gitArgs = []string{"diff", "HEAD"}
	}
	diff := safeGit(repoPath, "", gitArgs...)
	if diff == "" {
		return textResult("No differences found.")
	}
	return textResult(diff)
}

// gitGetDiffPaged applies the maxChars/charOffset paging from src/index.ts.
func gitGetDiffPaged(args map[string]any) toolResult {
	result := getDiff(args)
	raw := result.Content[0].Text
	offset := argInt(args, "charOffset")
	limit := argIntDefault(args, "maxChars", 8000)
	if offset == 0 && len(raw) <= limit {
		return result
	}
	if offset > len(raw) {
		offset = len(raw)
	}
	end := offset + limit
	if end > len(raw) {
		end = len(raw)
	}
	chunk := raw[offset:end]
	remaining := len(raw) - offset - len(chunk)
	suffix := ""
	if remaining > 0 {
		suffix = fmt.Sprintf("\n\n... (%d more chars, use charOffset=%d)", remaining, offset+len(chunk))
	}
	return textResult(chunk + suffix)
}

type remoteBranchInfo struct {
	exists  bool
	author  string
	date    string
	message string
	sha     string
}

func checkRemoteBranch(branchName, repoPath string) (remoteBranchInfo, error) {
	if err := validateBranch(branchName, "branchName"); err != nil {
		return remoteBranchInfo{}, err
	}
	lsRemote := safeGit(repoPath, "", "ls-remote", "--heads", "origin", "refs/heads/"+branchName)
	if lsRemote == "" {
		return remoteBranchInfo{exists: false}, nil
	}
	sha := strings.TrimSpace(strings.Fields(lsRemote)[0])
	shortSha := sha
	if len(shortSha) > 8 {
		shortSha = shortSha[:8]
	}
	if _, err := gitIn(repoPath, "fetch", "origin", branchName); err != nil {
		return remoteBranchInfo{exists: true, sha: shortSha}, nil
	}
	logLine := safeGit(repoPath, "", "log", "origin/"+branchName, "-1", "--format=%an%x09%ae%x09%ad%x09%s")
	if logLine == "" {
		return remoteBranchInfo{exists: true, sha: shortSha}, nil
	}
	parts := strings.SplitN(logLine, "\t", 4)
	info := remoteBranchInfo{exists: true, sha: shortSha}
	if len(parts) >= 1 {
		info.author = parts[0]
	}
	if len(parts) >= 2 && parts[1] != "" {
		info.author = parts[0] + " <" + parts[1] + ">"
	}
	if len(parts) >= 3 {
		info.date = parts[2]
	}
	if len(parts) >= 4 {
		info.message = parts[3]
	}
	return info, nil
}

func getDefaultBranch(repoPath string) string {
	head := safeGit(repoPath, "", "rev-parse", "--abbrev-ref", "origin/HEAD")
	if strings.HasPrefix(head, "origin/") {
		return strings.TrimPrefix(head, "origin/")
	}
	if safeGit(repoPath, "", "rev-parse", "--verify", "origin/main") != "" {
		return "main"
	}
	return "master"
}

func checkoutRemoteBranch(branchName, repoPath string) toolResult {
	if err := validateBranch(branchName, "branchName"); err != nil {
		return textResult("Error checking out branch: " + err.Error())
	}
	existing := safeGit(repoPath, "", "branch", "--list", branchName)
	if strings.TrimSpace(existing) != "" {
		if _, err := gitIn(repoPath, "checkout", branchName); err != nil {
			return textResult("Error checking out branch: " + err.Error())
		}
		return textResult(fmt.Sprintf("Switched to existing local branch %q.", branchName))
	}
	if _, err := gitIn(repoPath, "checkout", "--track", "origin/"+branchName); err != nil {
		return textResult("Error checking out branch: " + err.Error())
	}
	return textResult(fmt.Sprintf("Checked out %q tracking origin/%s.", branchName, branchName))
}

func createBranch(branchName, baseBranch, repoPath string, push bool) toolResult {
	if repoPath == "" {
		repoPath = mustGetwd()
	}
	if err := validateRepoPath(repoPath); err != nil {
		return textResult("Error creating branch: " + err.Error())
	}
	if !safeBranchRe.MatchString(branchName) {
		return textResult(fmt.Sprintf("Invalid branch name %q. Use only letters, numbers, /, _, ., -", branchName))
	}
	if baseBranch == "" {
		baseBranch = getDefaultBranch(repoPath)
	}
	if !safeBranchRe.MatchString(baseBranch) {
		return textResult(fmt.Sprintf("Invalid base branch name %q. Use only letters, numbers, /, _, ., -", baseBranch))
	}
	existing := safeGit(repoPath, "", "branch", "--list", branchName)
	if strings.TrimSpace(existing) != "" {
		return textResult(fmt.Sprintf("Branch %q already exists locally. Switch with: git checkout %s", branchName, branchName))
	}
	safeGit(repoPath, "", "fetch", "origin", baseBranch)
	if _, err := gitIn(repoPath, "checkout", "-b", branchName, "origin/"+baseBranch); err != nil {
		return textResult("Error creating branch: " + err.Error())
	}
	lines := []string{fmt.Sprintf("Created and switched to branch %q from origin/%s.", branchName, baseBranch)}
	if push {
		if _, err := gitIn(repoPath, "push", "-u", "origin", branchName); err != nil {
			return textResult("Error creating branch: " + err.Error())
		}
		lines = append(lines, fmt.Sprintf("Pushed to origin/%s and set upstream.", branchName))
	}
	return textResult(strings.Join(lines, "\n"))
}

func splitNonEmpty(s, sep string) []string {
	var out []string
	for _, p := range strings.Split(s, sep) {
		if p != "" {
			out = append(out, p)
		}
	}
	return out
}

func orValue(s, fallback string) string {
	if s == "" {
		return fallback
	}
	return s
}
